Microsoft.Quantum.Canon/StatePreparation/StatePreparation.qs

```diff
@@ -0,0 +1,208 @@
+    {
+        let nCoefficients = Length(coefficients);
+        mutable coefficientsComplexPolar = new ComplexPolar[nCoefficients];
+        for(idx in 0..nCoefficients - 1){
+            set coefficientsComplexPolar[idx] = ComplexPolar(AbsD(coefficients[idx]), 0.0);
+        }
+
+        return StatePreparationSBM(coefficientsComplexPolar, _);
+    }
+
+    /// # Summary
+    /// Returns a unitary operation $U$ that prepares an arbitrary quantum 
+    /// state $\ket{\psi}$ with complex coefficients $r_j e^{i t_j}$ from 
+    /// the $n$-qubit computational basis state $\ket{0...0}$.
+    ///
+    /// $U\ket{0...0}=\ket{\psi}=\frac{\sum^{2^n-1}_{j=0}r_j e^{i t_j}\ket{j}}{\sqrt{\sum^{2^n-1}_{j=0}|r_j|^2}}$.
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to $2^n$ complex coefficients represented by their 
+    /// absolute value and phase $(r_j, t_j)$. The $j$th coefficient 
+    /// indexes the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// # Output
+    /// A state-preparation unitary operation $U$.
+    ///
+    /// # Remarks
+    /// Negative input coefficients $r_j < 0$ will be treated as though
+    /// positive with value $|r_j|$. `coefficients` will be padded with 
+    /// elements $(r_j, t_j) = (0.0, 0.0)$ if fewer than $2^n$ are 
+    /// specified.
+    function StatePreparationComplexCoefficients (coefficients: ComplexPolar[]) : (BigEndian => (): Adjoint, Controlled)
+    {
+        return StatePreparationSBM(coefficients, _);
+    }
+
+    
+    /// # Summary
+    /// Returns a unitary operation $U$ that prepares an arbitrary quantum 
+    /// state $\ket{\psi}$ with complex coefficients $r_j e^{i t_j}$ from 
+    /// the $n$-qubit computational basis state $\ket{0...0}$.
+    ///
+    /// $U\ket{0...0}=\ket{\psi}=\frac{\sum^{2^n-1}_{j=0}r_j e^{i t_j}\ket{j}}{\sqrt{\sum^{2^n-1}_{j=0}|r_j|^2}}$.
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to $2^n$ complex coefficients represented by their 
+    /// absolute value and phase $(r_j, t_j)$. The $j$th coefficient 
+    /// indexes the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// ## qubits
+    /// Qubit register encoding number states in big-endian format. This is
+    /// expected to be initialized in the computational basis state 
+    /// $ket{0...0}$.
+    ///
+    /// # Remarks
+    /// Negative input coefficients $r_j < 0$ will be treated as though
+    /// positive with value $|r_j|$. `coefficients` will be padded with 
+    /// elements $(r_j, t_j) = (0.0, 0.0)$ if fewer than $2^n$ are 
+    /// specified.
+    ///
+    /// # References
+    /// - Synthesis of Quantum Logic Circuits
+    ///   Vivek V. Shende, Stephen S. Bullock, Igor L. Markov
+    ///   https://arxiv.org/abs/quant-ph/0406176
+    operation StatePreparationSBM (coefficients: ComplexPolar[], qubits: BigEndian) : ()
```

Comment:
Since this operation performs an action, I think it'd make sense to name it with a verb, such as `PrepareArbitraryState`. Esp. since we only have the one algorithm implemented, I'd also suggest dropping the `SBM` suffix. If we want to implement alternative algorithms, they should ideally be distinguished by an additional argument.

Reply:
Renamed to `PrepareArbitraryState`, removed `SBM`.